File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    void helperUtil(string op, int open, int close, vector<string>& res){
        
        if(open == 0 and close == 0){
            res.push_back(op);
            return;
        }    
        
        if(open != 0){
            helperUtil(op+"(", open-1, close, res);
        }
        if(close > open){
            helperUtil(op + ")", open, close-1, res);
        }
    }
    
    vector<string> generateParenthesis(int n) {
        
        vector<string> res;
        int open=n;
        int close=n;
        helperUtil("", open, close, res);
        return res;
    }
};

File: 22-generate-parentheses/22-generate-parentheses.py
class Solution:
    def helperUtil(self, output, opened, closed, res):
        if opened == 0 and closed == 0:
            res.append(output)
            return
        
        if opened != 0:
            self.helperUtil(output + "(", opened-1, closed, res)
        if closed > opened:
            self.helperUtil(output + ")", opened, closed-1, res)
        
        
    def generateParenthesis(self, n: int) -> List[str]:
        
        res = []
        opened, closed = n, n
        self.helperUtil("", opened, closed, res)
        return res